### ISSUE TITLE
map_script can cause Elasticsearch process to crash

### ISSUE THREAD

**Issue by n0othing** [elastic/elasticsearch#66175]
**Elasticsearch version** (`bin/elasticsearch --version`): `Version: 7.10.0, Build: default/tar/51e9d6f22758d0374a0f3f5c6e8f3a7997850f96/2020-11-09T21:30:33.964949Z, JVM: 15.0.1`

**Plugins installed**: []

**JVM version** (`java -version`): `JVM: 15.0.1`

**OS version** (`uname -a` if on a Unix-like system): `Darwin MacBook-Pro.localdomain 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64`

**Description of the problem including expected versus actual behavior**:

The below `map_script` causes the Elasticsearch process to crash/exit due to the `return` statement.

(Note: the `return` statement is not necessary-- the script will work fine if it's omitted).

**Steps to reproduce**:

```
GET myindex/_search
{
  "aggregations": {
    "myagg": {
      "scripted_metric": {
        "init_script": {
          "source": "state.mystring=''",
          "lang": "painless"
        },
        "map_script": {
          "source": "state.mystring= params._source.foo; return state.mystring;",
          "lang": "painless"
        },
        "combine_script": {
          "source": "return state.mystring;",
          "lang": "painless"
        },
        "reduce_script": {
          "source": "return states[0];",
          "lang": "painless"
        }
      }
    }
  }
}
```

**Provide logs (if relevant)**:

```
[2020-12-10T09:59:01,108][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [MacBook-Pro.localdomain] fatal error in thread [elasticsearch[MacBook-Pro.localdomain][search][T#11]], exiting
java.lang.NoClassDefFoundError: V
	at org.elasticsearch.painless.PainlessScript$Script.execute(state.mystring= params._source.foo; return state.mystring;:49) ~[?:?]
	at org.elasticsearch.search.aggregations.metrics.ScriptedMetricAggregator$1.collect(ScriptedMetricAggregator.java:135) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:107) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.apache.lucene.search.MultiCollector$MultiLeafCollector.collect(MultiCollector.java:192) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.apache.lucene.search.MatchAllDocsQuery$1$1.score(MatchAllDocsQuery.java:64) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:56) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:39) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:226) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:199) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:445) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
	at org.elasticsearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:341) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.query.QueryPhase.executeInternal(QueryPhase.java:296) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:148) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:372) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:431) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.SearchService.access$500(SearchService.java:141) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:401) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) ~[elasticsearch-7.10.0.jar:7.10.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.0.jar:7.10.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.ClassNotFoundException: V
	at java.lang.ClassLoader.findClass(ClassLoader.java:719) ~[?:?]
	at org.elasticsearch.painless.Compiler$Loader.findClass(Compiler.java:102) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	... 26 more
fatal error in thread [elasticsearch[MacBook-Pro.localdomain][search][T#11]], exiting
java.lang.NoClassDefFoundError: V
	at org.elasticsearch.painless.PainlessScript$Script.execute(state.mystring= params._source.foo; return state.mystring;:49)
	at org.elasticsearch.search.aggregations.metrics.ScriptedMetricAggregator$1.collect(ScriptedMetricAggregator.java:135)
	at org.elasticsearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:107)
	at org.apache.lucene.search.MultiCollector$MultiLeafCollector.collect(MultiCollector.java:192)
	at org.apache.lucene.search.MatchAllDocsQuery$1$1.score(MatchAllDocsQuery.java:64)
	at org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:56)
	at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:39)
	at org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:226)
	at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:199)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:445)
	at org.elasticsearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:341)
	at org.elasticsearch.search.query.QueryPhase.executeInternal(QueryPhase.java:296)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:148)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:372)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:431)
	at org.elasticsearch.search.SearchService.access$500(SearchService.java:141)
	at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:401)
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.ClassNotFoundException: V
	at java.base/java.lang.ClassLoader.findClass(ClassLoader.java:719)
	at org.elasticsearch.painless.Compiler$Loader.findClass(Compiler.java:102)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 26 more
```

**Comment by marcosvrrs** [comment]
Had this exact problem yesterday using the ingest pipeline with a return statement. One of my nodes crashed, a few hours later recovered (I'm using elastic cloud).

**Elasticsearch**:
"version" : "7.10.1", "build_flavor" : "default", "build_type" : "docker"

**Plugins installed**: []

**JVM**:
"version" : "15.0.1"

**OS** : 
"CentOS Linux 8 (Core)", "arch" : "amd64", "version" : "4.15.0-1039-aws",    "available_processors" : 2,"allocated_processors" : 2


I don't have the full log as I didn't have monitoring enabled, so got this partial log through the cloud console.

> [instance-0000000015] fatal error in thread [elasticsearch[instance-0000000015][management][T#4]], exiting java.lang.NoClassDefFoundError: V at org.elasticsearch.painless.PainlessScript$Script.execute( return ctx.velocidade_processamento = Instant.parse(ctx.process_start); :13) ~[?:?] at org.elasticsearch.ingest.common.ScriptProcessor.execute(ScriptProcessor.java:99) ~[?:?] at org.elasticsearch.ingest.Processor.execute(Processor.java:52) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.TrackingResultProcessor.execute(TrackingResultProcessor.java:104) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.innerExecute(CompoundProcessor.java:143) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.lambda$innerExecute$1(CompoundProcessor.java:162) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.innerExecute(CompoundProcessor.java:134) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.lambda$innerExecute$1(CompoundProcessor.java:162) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.TrackingResultProcessor.lambda$execute$1(TrackingResultProcessor.java:118) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.Processor.execute(Processor.java:57) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.TrackingResultProcessor.execute(TrackingResultProcessor.java:104) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.innerExecute(CompoundProcessor.java:143) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.execute(CompoundProcessor.java:129) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.innerExecute(CompoundProcessor.java:143) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.CompoundProcessor.execute(CompoundProcessor.java:129) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.Pipeline.execute(Pipeline.java:100) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.ingest.IngestDocument.executePipeline(IngestDocument.java:742) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.action.ingest.SimulateExecutionService.executeDocument(SimulateExecutionService.java:53) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.action.ingest.SimulateExecutionService.lambda$execute$3(SimulateExecutionService.java:82) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) ~[elasticsearch-7.10.1.jar:7.10.1] at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.1.jar:7.10.1] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?] at java.lang.Thread.run(Thread.java:832) [?:?] Caused by: java.lang.ClassNotFoundException: V at java.lang.ClassLoader.findClass(ClassLoader.java:719) ~[?:?] at org.elasticsearch.painless.Compiler$Loader.findClass(Compiler.java:102) ~[?:?] at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?] at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?] ... 25 more

**Comment by jdconrad** [comment]
I was able to replicate this.

**Comment by jdconrad** [comment]
This is fixed by #66957
